### PR TITLE
[FLINK-28098] Refactor table store compactor

### PR DESCRIPTION
**Pull request by JingsongLi**
There is currently some room for code optimization on the path of the compact.

- It uses the CompactManager, which actually only uses the CompactTask.
- It also doesn't need to use an asynchronous thread to complete the compaction.
- There should be FileStoreTable to provide unified table-level compaction.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/sink/TableCompact.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    private BiPredicate<BinaryRowData, Integer> partBucketFilter;
+
+    public TableCompact(FileStoreScan scan, FileStoreWrite<?> write, RowType partitionType) {
+        this.scan = scan;
+        this.write = write;
+        this.partitionType = partitionType;
+    }
+
+    public TableCompact withPartitions(Map<String, String> partitionSpec) {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/sink/TableCompact.java:58]:
```suggestion
    public TableCompact withPartition(Map<String, String> partitionSpec) {
```
Nit: actually there's just one partition can be specified

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AppendOnlyFileStoreWrite.java [hunk 2]
```diff
@@ -71,11 +75,8 @@ public RecordWriter<RowData> createEmptyWriter(
     }
 
     @Override
-    public RecordWriter<RowData> createCompactWriter(
-            BinaryRowData partition,
-            int bucket,
-            ExecutorService compactExecutor,
-            List<DataFileMeta> restoredFiles) {
+    public Callable<CompactResult> createCompactWriter(
+            BinaryRowData partition, int bucket, @Nullable List<DataFileMeta> compactFiles) {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AppendOnlyFileStoreWrite.java:79]: Just curious about which condition `compactFiles` is passed as `null`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreWrite.java [hunk 3]
```diff
@@ -40,10 +44,12 @@ RecordWriter<T> createWriter(
...
-            ExecutorService compactExecutor,
-            List<DataFileMeta> restoredFiles);
+    /**
+     * Create a {@link Callable} compactor from partition, bucket.
+     *
+     * @param compactFiles input files of compaction. When it is null, will automatically read all
+     *     files of the current bucket.
+     */
+    Callable<CompactResult> createCompactWriter(
+            BinaryRowData partition, int bucket, @Nullable List<DataFileMeta> compactFiles);
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreWrite.java:54]: Just curious about which condition `compactFiles` is passed as `null`

**Review by LadyForest — COMMENTED** [review]
Thanks for the contribution. Looks pretty good, I just left some minor comments.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/sink/TableCompact.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/table/sink/TableCompact.java:58]: `partitionSpec` can be partial, which means there can be many partitions.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreWrite.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreWrite.java:54]: Actually, here we can just pass a null, but for the performance, we reuse compact files.

**Review by LadyForest — APPROVED** [review]
Thanks for the explanation. Looks good to me :)

**Pull request merged**